### PR TITLE
Implement directory support for wf-background

### PR DESCRIPTION
**Pull request by soreau**
Allows users to set the background image key to a directory.

### CODE REVIEW

**Comment by soreau** [comment]
https://streamable.com/yeg88

File: src/background/background.cpp [hunk 1]
```diff
@@ -1,3 +1,7 @@
+#include <sys/stat.h>
+#include <wordexp.h>
+#include <dirent.h>
+#include <gtkmm.h>
```
Comment (ammen99) [inline src/background/background.cpp]: I try to include only the necessary files. Are they too much?

File: src/background/background.cpp [hunk 2]
```diff
@@ -9,117 +13,312 @@
...
-    zwf_wm_surface_v1 *wm_surface = NULL;
+    auto pbuf = background->pbuf;
+    auto pbuf2 = background->pbuf2;
+    auto animation = &background->fade_animation;
+    auto alpha = animation->progress();
+    auto images = background->images;
 
-    Gtk::Window window;
-    Gtk::Image image;
+    if (animation->running())
```
Comment (ammen99) [inline src/background/background.cpp]: Braces {} for if/else. No braces are allowed only when both those things happen:
1. No else part
2. Single-line statement

The code isn't updated everywhere I agree, but this is the style that should be used.
Which reminds me I should create a style guide somewhere ...

File: src/background/background.hpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+{
+    WayfireShellApp *app;
+    WayfireOutput *output;
+    zwf_wm_surface_v1 *wm_surface = NULL;
+
+    BackgroundDrawingArea drawing_area;
+    Gtk::Window window;
+    Gtk::Image image;
+    int scale;
+    int cycle_timeout;
```
Comment (ammen99) [inline src/background/background.hpp]: Is there a reason why we can't use `background_cycle_timeout->as_int()` directly?

File: src/background/background.hpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    bool inhibited = true;
+    uint background[2] = {0};
```
Comment (ammen99) [inline src/background/background.hpp]: Maybe remain this to `active_background_indices`?

File: src/background/background.cpp [hunk 5]
```diff
@@ -9,117 +13,312 @@
...
+    pbuf = create_from_file_safe(path);
+
+    if (!load_next_background(path, background[1]))
+        return false;
+
+    std::cout << "Loaded " << path << std::endl;
+
+    pbuf2 = create_from_file_safe(images[background[1]]);
+
+    pbuf = pbuf->scale_simple(output_width * scale,
```
Comment (ammen99) [inline src/background/background.cpp]: Style issue: multi-line function calls start on the next line with 4 spaces identation, not aligned to the opening (

File: src/background/background.cpp [hunk 6]
```diff
@@ -9,117 +13,312 @@
...
+            continue;
+
+        auto fullpath = std::string(exp.we_wordv[0]) + "/" + file->d_name;
+
+        struct stat next;
+        if (stat(fullpath.c_str(), &next) == 0)
         {
-            std::cerr << "Error: created window was not a wayland surface" << std::endl;
-            std::exit(-1);
+            if (S_ISDIR(next.st_mode))
```
Comment (ammen99) [inline src/background/background.cpp]: Braces for the if/else

File: src/background/background.cpp [hunk 7]
```diff
@@ -9,117 +13,312 @@
...
-            }
-        } catch (...)
+void
+WayfireBackground::reset_background()
+{
+    fade_animation.start(0.0, 0.0);
+    images.clear();
+    change_bg_conn.disconnect();
+    background[0] = background[1] = 0;
+    if (pbuf)
```
Comment (ammen99) [inline src/background/background.cpp]: braces {}

File: src/background/background.cpp [hunk 8]
```diff
@@ -9,117 +13,312 @@
...
+                throw std::exception();
         }
-    }
 
-    void setup_window()
-    {
-        window.set_resizable(false);
-        window.set_decorated(false);
+        scale = window.get_scale_factor();
+        pbuf = pbuf->scale_simple(output_width * scale,
```
Comment (ammen99) [inline src/background/background.cpp]: identation: 4 spaces instead of align with (

File: src/background/background.cpp [hunk 9]
```diff
@@ -9,117 +13,312 @@
...
+        }
 
-        window.property_scale_factor().signal_changed().connect(
-            sigc::mem_fun(this, &WayfireBackground::set_background));
+        fade_animation.start(0.0, 1.0);
+        drawing_area.queue_draw();
+    } catch (...)
+    {
+        window.remove();
+        if (images.size())
```
Comment (ammen99) [inline src/background/background.cpp]: use {}

File: src/background/background.hpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+class WayfireBackground;
+
+class BackgroundDrawingArea : public Gtk::DrawingArea
+{
+    WayfireBackground *background;
+
+    public:
+    BackgroundDrawingArea(WayfireBackground *background);
```
Comment (ammen99) [inline src/background/background.hpp]: I still don't think we should pass the whole `WayfireBackground` to the drawing area. Here is a design which I think should be not-so-hard to implement (correct me if I'm wrong though):

1. Have fade animation managed completely internally by the DrawingArea (so nothing fade-related in WayfireBackground)
2. Have a method like `show_image(Glib::RefPtr<Gdk::PixBuf> next_image)` to the drawing area. What it does is basically tell the drawing area to fade from the current image to the `new_image`.

This way we'd have the actual rendering/fading in the DrawingArea, while Background manages what is to be shown. You could then store `pbuf`, `pbuf2` etc in DrawingArea.

Comment (soreau) [reply src/background/background.hpp]: It took a bit of back and forth but I think I managed to work it out like you described.

File: src/background/background.cpp (hunk 9 above)
Comment (soreau) [reply src/background/background.cpp]: Yep.

File: src/background/background.cpp (hunk 8 above)
Comment (soreau) [reply src/background/background.cpp]: Done.

File: src/background/background.cpp (hunk 7 above)
Comment (soreau) [reply src/background/background.cpp]: Actually this part falls away with your suggestion below.

File: src/background/background.cpp (hunk 6 above)
Comment (soreau) [reply src/background/background.cpp]: Check.

File: src/background/background.cpp (hunk 5 above)
Comment (soreau) [reply src/background/background.cpp]: This is how it was before I touched it, will change all instances.

File: src/background/background.hpp (hunk 4 above)
Comment (soreau) [reply src/background/background.hpp]: That seems a bit long, how about just active_background? Going with this for now.

File: src/background/background.hpp (hunk 3 above)
Comment (soreau) [reply src/background/background.hpp]: It's a convenience variable but I moved it out of the header and just made them local variables, if that's ok with you.

File: src/background/background.cpp (hunk 2 above)
Comment (soreau) [reply src/background/background.cpp]: I have been just trying to use braces for both `if` and `else`, when one or the other has more than one line, or none at all when both have a single line, but will use braces from now on unless it's a single `if` followed by single line.

File: src/background/background.cpp (hunk 1 above)
Comment (soreau) [reply src/background/background.cpp]: If I try to remove gtkmm.h, I get `error: ‘signal_timeout’ is not a member of ‘Glib’` and wordexp.h is also required. I removed sys/stat.h and dirent.h however.

Comment (ammen99) [reply src/background/background.cpp]: The header for `signal_timeout` is `#include <glibmm/main.h>` according to https://developer.gnome.org/glibmm/stable/classGlib_1_1SignalTimeout.html

File: src/background/background.hpp [hunk 11]
```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+class WayfireBackground;
+
+class BackgroundDrawingArea : public Gtk::DrawingArea
+{
+    wf_duration fade;
+    Glib::RefPtr<Gdk::Pixbuf> pbuf, pbuf2;
```
Comment (ammen99) [inline src/background/background.hpp]: Could we use a bit more descriptive names? Maybe current_image and last_image or something else you find better?

File: src/background/background.hpp [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+class WayfireBackground
+{
+    WayfireShellApp *app;
+    WayfireOutput *output;
+    zwf_wm_surface_v1 *wm_surface = NULL;
+
+    BackgroundDrawingArea drawing_area;
+    std::vector<std::string> images;
+    Gtk::Window window;
+    Gtk::Image image;
```
Comment (ammen99) [inline src/background/background.hpp]: I feel like this isn't needed anymore ...

File: src/background/background.cpp [hunk 13]
```diff
@@ -9,117 +11,305 @@
...
+    }
+    return true;
+}
+
+bool
+WayfireBackground::load_next_background(Glib::RefPtr<Gdk::Pixbuf> &pbuf, std::string &path)
+{
+    while (!pbuf)
+    {
+        if (++current_background > images.size() - 1)
```
Comment (ammen99) [inline src/background/background.cpp]: one-liner: `current_background = (current_background + 1) % images.size()`

File: src/background/background.cpp [hunk 14]
```diff
@@ -9,117 +11,305 @@
...
+            current_background = 0;
+
+        path = images[current_background];
+        pbuf = create_from_file_safe(path);
+
+        if (!pbuf)
+            images.erase(images.begin() + current_background);
+
+        if (!images.size())
```
Comment (ammen99) [inline src/background/background.cpp]: Shouldn't this be in the very start of the loop body? This way the function also handles empty images vector.

File: src/background/background.cpp (hunk 1 above)
Comment (soreau) [reply src/background/background.cpp]: That might be, but if I try to remove gtkmm.h, I get a whole slew of errors. Do you want me to figure out exactly what headers are required or why gtkmm.h is necessary to include?

File: src/background/background.hpp (hunk 11 above)
Comment (soreau) [reply src/background/background.hpp]: Would to_pbuf and from_pbuf work?

File: src/background/background.hpp (hunk 12 above)
Comment (soreau) [reply src/background/background.hpp]: Removed.

File: src/background/background.cpp (hunk 13 above)
Comment (soreau) [reply src/background/background.cpp]: Great.

File: src/background/background.cpp (hunk 14 above)
Comment (soreau) [reply src/background/background.cpp]: Yes, you're right.

File: src/background/background.cpp (hunk 1 above)
Comment (soreau) [reply src/background/background.cpp]: Actually, now I've whittled it down to `error: ‘Gdk::Cairo’ has not been declared` but can't seem to figure out which header is needed.

Comment (soreau) [reply src/background/background.cpp]: It works if I include glibmm/main.h, gtkmm/drawingarea.h and gdkmm/general.h.

File: src/background/background.hpp (hunk 11 above)
Comment (ammen99) [reply src/background/background.hpp]: Yep. Maybe it is best to add a comment to describe them better.

**Comment by soreau** [comment]
What is the status of this PR?

**Comment by ammen99** [comment]
I've forgotten about this. There don't seem to be any issues so ...

**Pull request merged**